### DISCUSSION TITLE
Support labels for agents and organization

### DISCUSSION THREAD

**Discussion by nedrebo** [community/community#10969] [Actions]
# Proposal

## Part 1

In addition to setting labels on runners, it should also be possible to set labels on an agent and on the organization level.

## Part 2

It should also be possible to make labels a constrained resource, not only an infinite resource like it is today. This should be supported for labels on runner, agent, and organization level.

# User stories

## Physically connected HW

Let's say you have an agent with 16 runners and you have 4 USB web cameras[^1] connected to the agent. The cameras are needed for testing, some tests use 1 camera some up to 4 cameras at the same time.

With today's setup options you would have to assign all 4 cameras to one single runner to be able to run the 4x camera test. This would cause queues if that runner was scheduled to do a non-camera job even if other runners on the same agent are free. And, if the runner was scheduled with a 1x camera job, then 3 cameras would be idle and not usable from other runners.

However, if it was possible to configure the _agent_ with 4 x `webcam` labels, then any job that needed cameras could be scheduled to any runner as long as cameras are available, and it would only lock up exactly the number of labels that the job asks for.

These snippets are quick examples of how this can be expressed in a workflow:

```yml
job-that-tests-two-cameras:
    runs-on: ["camera", "camera", "gpu-amd"] # Note the duplicated label
```

```yml
job-that-tests-two-cameras:
    runs-on: ["camera:2", "gpu-amd"]
```

The last one may not be possible if `camera:2` is a valid label today. But I think the idea should be clear and you'll find a good grammar.

## Network resources

This is very much the same scenario as above except it covers network resources. Let's now imagine you have 4 cameras that are available over mDNS. That means you can run 1-4 camera jobs in the CI depending on how many cameras each job need. However, the difference now is that any runner on any agent should be able to process such jobs to secure max CI throughput. There's no reason to tie these cameras to any single runner or single agent, that would just be limiting.

# Why not use a 3rd party job queue system

This is what I imagine many people are doing to work around this. It's even been suggested to be done at our place. It is doable, but it has the following flaws:

* It is not simplicity. I already have an up and running queue system, GitHub Actions itself. It should handle it.
* The workflow state will no longer be reflected correctly in the GitHub UI. All jobs in the workflow will appear to do work because the Actions UI is not aware of the other queue system. In fact, your workflow could be stuck completely due to cameras being down while all jobs are appearing to be active in the UI. This will delay feedback and cause errors to go undetected for longer.

The latter is the bigger concern. I could live with the investment of setting up a second queue system, but when the developer experience is also going to take a hit, then that is not a tempting option. So I really hope this can be supported by GitHub Actions.

---
This feature request relates to https://github.com/github/feedback/discussions/10178. Together they will make the job scheduling in GitHub Actions smooth and cover all our current use cases.

---
[^1]: I have used webcams as an example, but this could be anything really: In our case: relays, GPUs, and the network and USB 3D sensors that our company builds and needs to be available in the CI.

**Comment by ethomson** [comment]
I once again appreciate your thoughtful feedback.  I think that I'm understanding your use case, but this one is a little tougher for me to see as a common problem.  🤔  I think that your "webcam" example is a good one - I'll reach out to some other people who are doing hardware testing with Actions and see if there's a more (or less) general problem that we can solve here.

**Reply by nedrebo** [reply]
Thx for the quick answer!

I can sell this a bit more :) This is also going to be a simplification even for the simplest case.

I imagine 99% of users have the same labels on all runners on the same agent. We at least do for all the "soft" labels. Having to repeat the label N times is tedious. Much (a tiny bit at least) simpler if one could just add the label once and for all to the agent.
